### PR TITLE
Deferred loading causes flickering with SSR

### LINKED ISSUE #9677: Deferred loading causes flickering with SSR
When deferred loading is used, there's a flickering issue caused when the page is initially served by SSR. Since the SSR generated page will ignore deferred loading, all DOM will land initially in the browser. When, however, the CSR kicks in, the DOM will be replaced by the CSR rendered DOM. The CSR app will respect the deferred loading configuration, which leads to some elements being cleared initially. 

We must avoid flickering to ensure a seamless transition from SSR to CSR.

### PR DESCRIPTION
**Pull request by dunqan**
Closes #9677

### CODE REVIEW

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  providedIn: 'root',
+})
+export class PageSlotService {
+  protected instantSsrSlots: string[] | undefined;
+
+  constructor(
+    @Inject(PLATFORM_ID) platformId: any,
+    @Inject(DOCUMENT) document
+  ) {
+    if (isPlatformBrowser(platformId)) {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: Let's avoid logic in the constructor to not block extensibily.

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        .filter(
+          (el: Element) =>
+            el.getBoundingClientRect().top <
+            document.documentElement.clientHeight
+        )
+        .map((el: Element) => el.classList[0]);
+    }
+  }
+
+  shouldNotDefer(slot: string): boolean {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:45]: what about `isPrerendered()`?

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      // Find all slots visible in the SSR pre-rendered DOM
+      this.instantSsrSlots = Array.from(
+        document.querySelectorAll('cx-page-slot')
+      )
+        .filter(
+          (el: Element) =>
+            el.getBoundingClientRect().top <
+            document.documentElement.clientHeight
+        )
+        .map((el: Element) => el.classList[0]);
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: This is a bit tricky. I'd consider introducing a new page slot attribute that indicates the page slot position.

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 2 above)
Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:45]: actually.. going forward, we probably want to move the component logic to this service. We can take a first stab at it by renaming this method, and implement all the logic of `PageSlotComponent.getComponentDeferOptions()`. That will make this API better readable (we hide the SSR complexity from the API).

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 1 above)
Comment (dunqan) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: 👍

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 3 above)
Comment (dunqan) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: 👍

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 2 above)
Comment (dunqan) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:45]: 👍

File: docs/migration/3_0.md [hunk 4]
```diff
@@ -183,3 +183,4 @@ to:
 - `ReviewSubmitComponent` no longer uses `CheckoutConfigService`. This service usage was replaced with the corresponding methods form `CheckoutStepService`. In addition, `PaymentTypeService`, `CheckoutCostCenterService` and `UserCostCenterService` need to be provided to `ReviewSubmitComponent`.
 - `ShippingAddressSetGuard` no longer uses `CheckoutConfigService`. This service usage was replaced with the corresponding methods form `CheckoutStepService`. This service needs to be provided to `ShippingAddressSetGuard`.
 - `ShippingAddressComponent` no longer uses `CheckoutConfigService` and `RoutingService`. These services usage was replaced with the corresponding methods form `CheckoutStepService`. This service needs to be provided to `ShippingAddressComponent`.
+- `PageSlotComponent` no loger uses `CmsComponentsService`. This servei usage was replaces with the `PageSlotService`.
```
Comment (tobi-or-not-tobi) [inline docs/migration/3_0.md]: typo

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    @Inject(PLATFORM_ID) protected platformId: any,
+    @Inject(DOCUMENT) protected document
+  ) {
+    this.discoverInstantSsrSlots();
+  }
+
+  /**
+   * Finds all slots visible in the SSR pre-rendered DOM
+   */
+  protected discoverInstantSsrSlots() {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: what about `resolvePrerenderedSlots()`

Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: should `void`

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 2 above)
Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:45]: Is this method still intended to be part of public API?

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  shouldNotDefer(slot: string): boolean {
+    if (this.instantSsrSlots?.includes(slot)) {
+      this.instantSsrSlots.splice(this.instantSsrSlots.indexOf(slot), 1);
+      return true;
+    }
+    return false;
+  }
+
+  getComponentDeferOptions(
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:57]: let's add a small doc to this public method,i.e.

> Returns the defer options for the given component. If the wrapping page slot is prerendered, we would ignore the defer options altogether.

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 2 above)
Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts:45]: nit: what about renaming, I don't like the `Not` part of it. Either `shouldDefer()` or `isPrerendered`.

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts (hunk 5 above)
Comment (dunqan) [reply projects/storefrontlib/src/cms-structure/page/slot/page-slot.service.ts]: 👍

File: docs/migration/3_0.md (hunk 4 above)
Comment (dunqan) [reply docs/migration/3_0.md]: 👍

**Review by tobi-or-not-tobi — APPROVED** [review]

**Pull request merged**